File: common/logisticspipes/utils/OpenGLDebugger.java
package logisticspipes.utils;

import java.awt.Color;
import java.awt.Component;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.event.KeyEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import javax.swing.BorderFactory;
import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JDialog;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTable;
import javax.swing.JTextField;
import javax.swing.KeyStroke;
import javax.swing.ScrollPaneConstants;
import javax.swing.UIManager;
import javax.swing.UnsupportedLookAndFeelException;
import javax.swing.WindowConstants;
import javax.swing.table.DefaultTableCellRenderer;
import javax.swing.table.DefaultTableModel;
import javax.swing.table.TableCellRenderer;

import lombok.Getter;
import lombok.Setter;
import org.lwjgl.opengl.GL11;

@SuppressWarnings("unused")
public class OpenGLDebugger {

	private static HashMap<Integer, String> niceToHave = null;
	private static int probeID = 0;
	private Thread probeGUIThread;
	private int cycleCount;
	private boolean started;
	private ExtendedHashMap glStuff;
	private ConcurrentHashMap<Integer, GLTypes> glVariablesToCheck;
	private final Lock debuggerLock;
	private final Condition glVariablesCondition;
	private boolean glVariablesUpdated;
	@Getter
	@Setter
	private int printOnCycle;

	private enum GLTypes {
		BOOLEAN(Boolean.class, "boolean", "GL11.glGetBoolean"),
		DOUBLE(Double.class, "double", "GL11.glGetDouble"),
		FLOAT(Float.class, "float", "GL11.glGetFloat"),
		INTEGER(Integer.class, "int", "GL11.glGetInteger"),
		INTEGER64(Long.class, "long", "GL32.glGetInteger64");

		private Class javaClass;
		private String getterFunction;
		private String niceName;

		GLTypes(Class javaClass, String niceName, String getterFunction) {
			this.javaClass = javaClass;
			this.niceName = niceName;
			this.getterFunction = getterFunction;
		}

		public Class getJavaClass() {
			return javaClass;
		}

		public String getGetterFunction() {
			return getterFunction;
		}

		public String getNiceName() {
			return niceName;
		}
	}

	public static class ExtendedHashMap extends HashMap<Integer, Object> {

		private ArrayList<Integer> orderedKeys;
		private ArrayList<Integer> newKeys;
		private ArrayList<Integer> updatedKeys;
		private boolean sessionStarted;

		public int getStopUpdatedIndex() {
			return stopUpdatedIndex;
		}

		public int getStopNewIndex() {
			return stopNewIndex;
		}

		private int stopUpdatedIndex;
		private int stopNewIndex;

		public ExtendedHashMap() {
			sessionStarted = false;
			orderedKeys = new ArrayList<>();
		}

		@Override
		public void putAll(Map<? extends Integer, ?> m) {
			throw new UnsupportedOperationException();
		}

		public void startSession() {
			newKeys = new ArrayList<>();
			updatedKeys = new ArrayList<>();
			sessionStarted = true;
		}

		@Override
		public Object put(Integer key, Object value) {
			if (!sessionStarted) {
				throw new UnsupportedOperationException("Session not started");
			}
			if (containsKey(key)) {
				if (get(key).equals(value)) {
					return value;
				} else {
					orderedKeys.remove(key);
					updatedKeys.add(key);
				}
			} else {
				newKeys.add(key);
			}
			return super.put(key, value);
		}

		public void stopSession() {
			stopNewIndex = newKeys.size();
			orderedKeys.addAll(0, newKeys);
			newKeys = null;
			stopUpdatedIndex = updatedKeys.size();
			stopNewIndex += stopUpdatedIndex;
			orderedKeys.addAll(0, updatedKeys);
			updatedKeys = null;
			sessionStarted = false;
		}

		public String getName(Integer key) {
			return OpenGLDebugger.niceToHave.get(key);
		}

		public int getKey(int index) {
			return orderedKeys.get(index);
		}
	}

	public class SpecialTableModel extends DefaultTableModel {

		@Override
		public boolean isCellEditable(int row, int column) {
			return false;
		}

		@Override
		public String getColumnName(int column) {
			switch (column) {
				case 0:
					return "Key";
				case 1:
					return "Value";
				default:
					return "<NOVALUE>";
			}
		}

		@Override
		public int getRowCount() {
			return glStuff.size();
		}

		@Override
		public int getColumnCount() {
			return 2;
		}

		@Override
		public Object getValueAt(int rowIndex, int columnIndex) {
			try {
				int index = glStuff.getKey(rowIndex);
				switch (columnIndex) {
					case 0:
						return glStuff.getName(index);
					case 1:
						return glStuff.get(index);
					default:
						return "<NOVALUE>";
				}
			} catch (IndexOutOfBoundsException e) {
				return "<EXCEPTION>";
			}
		}
	}

	public class SpecialTableCellRenderer extends DefaultTableCellRenderer {

		@Override
		public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column) {
			if (table == null) {
				return this;
			}
			setBackground(null);
			super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column);

			if (row < glStuff.getStopUpdatedIndex()) {
				setBackground(Color.YELLOW);
			} else if (row < glStuff.getStopNewIndex()) {
				setBackground(Color.GREEN);
			}

			return this;
		}
	}

	public class ProbeGUI extends JDialog implements Runnable {

		private JPanel mainPanel;
		private JTable variableMonitorTable;
		private JButton closeButton;

		public ProbeGUI() {
			for (UIManager.LookAndFeelInfo info : UIManager.getInstalledLookAndFeels()) {
				if ("Windows".equals(info.getName())) {
					try {
						UIManager.setLookAndFeel(info.getClassName());
					} catch (ReflectiveOperationException | UnsupportedLookAndFeelException e) {
						e.printStackTrace();
					}
					break;
				}
			}

			setupUI();

			setType(Type.UTILITY);
			setContentPane(mainPanel);
			getRootPane().setDefaultButton(closeButton);

			variableMonitorTable.setModel(new SpecialTableModel());

			TableCellRenderer cellRenderer = new SpecialTableCellRenderer();
			variableMonitorTable.getColumnModel().getColumn(0).setCellRenderer(cellRenderer);

			setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
			addWindowListener(new WindowAdapter() {

				@Override
				public void windowClosing(WindowEvent e) {
					stop();
				}
			});

			mainPanel.registerKeyboardAction(e -> stop(), KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT);
		}

		@Override
		public void run() {
			for (Integer key : OpenGLDebugger.niceToHave.keySet()) {
				glVariablesToCheck.put(key, GLTypes.BOOLEAN);
			}
			pack();
			setVisible(true);

			while (started) {
				debuggerLock.lock();
				try {
					while (!glVariablesUpdated) {
						glVariablesCondition.await();
					}
					glVariablesUpdated = false;
					updateVariables();
				} catch (InterruptedException e) {
					e.printStackTrace();
				} finally {
					debuggerLock.unlock();
				}
			}
		}

		private void updateVariables() {
			DefaultTableModel dtm = (DefaultTableModel) variableMonitorTable.getModel();
			dtm.fireTableDataChanged();
		}

		private void setupUI() {
			mainPanel = new JPanel();
			mainPanel.setLayout(new GridBagLayout());
			mainPanel.setBorder(BorderFactory.createTitledBorder(BorderFactory.createEmptyBorder(5, 5, 5, 5), null));
			closeButton = new JButton();
			closeButton.setText("Close");
			GridBagConstraints gbc;
			gbc = new GridBagConstraints();
			gbc.gridx = 2;
			gbc.gridy = 1;
			gbc.fill = GridBagConstraints.HORIZONTAL;
			gbc.insets = new Insets(0, 5, 0, 0);
			mainPanel.add(closeButton, gbc);
			JButton addButton = new JButton();
			addButton.setText("Add");
			gbc = new GridBagConstraints();
			gbc.gridx = 1;
			gbc.gridy = 1;
			gbc.fill = GridBagConstraints.HORIZONTAL;
			gbc.insets = new Insets(0, 5, 0, 0);
			mainPanel.add(addButton, gbc);
			JTextField addTextField = new JTextField();
			addTextField.setText("");
			gbc = new GridBagConstraints();
			gbc.gridx = 0;
			gbc.gridy = 1;
			gbc.weightx = 1.0;
			gbc.anchor = GridBagConstraints.WEST;
			gbc.fill = GridBagConstraints.HORIZONTAL;
			mainPanel.add(addTextField, gbc);
			JScrollPane monitorTableScrollPane = new JScrollPane();
			monitorTableScrollPane.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER);
			gbc = new GridBagConstraints();
			gbc.gridx = 0;
			gbc.gridy = 0;
			gbc.gridwidth = 3;
			gbc.weightx = 1.0;
			gbc.weighty = 1.0;
			gbc.fill = GridBagConstraints.BOTH;
			gbc.insets = new Insets(0, 0, 5, 0);
			mainPanel.add(monitorTableScrollPane, gbc);
			variableMonitorTable = new JTable();
			variableMonitorTable.setEnabled(false);
			monitorTableScrollPane.setViewportView(variableMonitorTable);
		}
	}

	public OpenGLDebugger(int printOnCycle) {
		if (printOnCycle < 1) {
			throw new IllegalArgumentException("Print per cycle must be at least 1");
		}

		if (OpenGLDebugger.niceToHave == null) {
			OpenGLDebugger.updateNiceToHave();
		}

		debuggerLock = new ReentrantLock();
		glVariablesCondition = debuggerLock.newCondition();

		this.printOnCycle = printOnCycle;
		glStuff = new ExtendedHashMap();
		glVariablesToCheck = new ConcurrentHashMap<>();

		probeGUIThread = new Thread(new ProbeGUI(), "LogisticsPipes GLDebug Probe #" + OpenGLDebugger.probeID);
		OpenGLDebugger.probeID++;
	}

	public void start() {
		if (!started) {
			started = true;
			cycleCount = 0;
			probeGUIThread.start();
		}
	}

	public void stop() {
		if (started) {
			debuggerLock.lock();
			try {
				started = false;
				glVariablesUpdated = true;
				glVariablesCondition.signal();
			} finally {
				debuggerLock.unlock();
			}
		}
	}

	public void cycle() {
		if (started) {
			++cycleCount;
			if (cycleCount % printOnCycle == 0) {
				saveOpenGLStuff();
				cycleCount = 0;
			}
		}
	}

	private void saveOpenGLStuff() {
		debuggerLock.lock();
		try {
			Iterator<Integer> i = glVariablesToCheck.keySet().iterator();
			glStuff.startSession();
			while (i.hasNext()) {
				Integer key = i.next();
				Object value = GL11.glGetBoolean(key);
				if (GL11.glGetError() == GL11.GL_INVALID_ENUM) {
					i.remove();
				} else {
					glStuff.put(key, value);
				}
			}
			glStuff.stopSession();
			glVariablesUpdated = true;
			glVariablesCondition.signal();
		} finally {
			debuggerLock.unlock();
		}
	}

	private static void updateNiceToHave() {
		OpenGLDebugger.niceToHave = new HashMap<>();
		int crawlerVersion = 11;
		boolean almostEnd = false;
		boolean end = false;
		while (!end) {
			String packageGL = String.format("%s%d", "GL", crawlerVersion);
			String nextGL = String.format("%s.%s", "org.lwjgl.opengl", packageGL);
			try {
				crawlerVersion++;
				Class glClass = GL11.class.getClassLoader().loadClass(nextGL);
				com.google.common.reflect.Reflection.initialize(glClass);
				almostEnd = false;

				for (Field f : glClass.getDeclaredFields()) {
					try {
						if (!f.getType().equals(int.class)) {
							continue;
						}

						int id = f.getInt(null);
						String nice = f.getName();
						if (nice.endsWith("BIT")) {
							continue;
						}

						// All the things that are being replaced are not that bad
						if (OpenGLDebugger.niceToHave.containsKey(id) && !OpenGLDebugger.niceToHave.get(id).equals(nice)) {
							System.out.printf("NiceToHave: ID %d exists. Replacing %s with %s!!%n", id, OpenGLDebugger.niceToHave.remove(id), nice);
						}

						OpenGLDebugger.niceToHave.put(id, String.format("%s.%s", packageGL, nice));
					} catch (IllegalArgumentException e) {
						System.out.printf("NiceToHave: Illegal Argument!%nNiceToHave: %s%n", e);
						e.printStackTrace();
					} catch (IllegalAccessException e) {
						System.out.printf("NiceToHave: Illegal Access!%nNiceToHave: %s%n", e);
						e.printStackTrace();
					}
				}
			} catch (ClassNotFoundException e) {
				if (almostEnd) {
					end = true;
				} else {
					almostEnd = true;
					crawlerVersion = (crawlerVersion / 10 + 1) * 10;
				}
			}
		}
	}
}

File: src/main/kotlin/network/rs485/debug/OpenGLDebugger.kt
/*
 * Copyright (c) 2022  RS485
 *
 * "LogisticsPipes" is distributed under the terms of the Minecraft Mod Public
 * License 1.0.1, or MMPL. Please check the contents of the license located in
 * https://github.com/RS485/LogisticsPipes/blob/dev/LICENSE.md
 *
 * This file can instead be distributed under the license terms of the
 * MIT license:
 *
 * Copyright (c) 2022  RS485
 *
 * This MIT license was reworded to only match this file. If you use the regular
 * MIT license in your project, replace this copyright notice (this line and any
 * lines below and NOT the copyright line above) with the lines from the original
 * MIT license located here: http://opensource.org/licenses/MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this file and associated documentation files (the "Source Code"), to deal in
 * the Source Code without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Source Code, and to permit persons to whom the Source Code is furnished
 * to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Source Code, which also can be
 * distributed under the MIT.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package network.rs485.debug

import com.google.common.reflect.Reflection
import logisticspipes.LogisticsPipes
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL31
import org.lwjgl.opengl.GL32
import org.lwjgl.util.glu.GLU
import java.awt.*
import java.awt.event.KeyEvent
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.util.*
import java.util.concurrent.locks.ReentrantLock
import javax.swing.*
import javax.swing.table.DefaultTableCellRenderer
import javax.swing.table.DefaultTableModel
import kotlin.concurrent.withLock

private class ChangeTrackingSortedMap<K : Comparable<K>, V : Any> {
    private val orderedKeys = ArrayList<K>()
    private val values = ArrayList<V>()
    private val sessionUpdates = ArrayList<Int>()
    private val sessionInsertions = ArrayList<Int>()
    private var sessionStarted: Boolean = false

    val size: Int
        get() = orderedKeys.size

    fun startSession() {
        sessionUpdates.clear()
        sessionInsertions.clear()
        sessionStarted = true
    }

    fun put(key: K, value: V) {
        if (!sessionStarted) {
            throw IllegalStateException("Session not started. Only modify map after session start.")
        }
        val keyIndex = binarySearch(key)
        if (keyIndex < 0) {
            val addIndex = -(keyIndex + 1)
            orderedKeys.add(addIndex, key)
            values.add(addIndex, value)
            sessionInsertions.add(addIndex)
        } else if (values[keyIndex] != value) {
            orderedKeys[keyIndex] = key
            values[keyIndex] = value
            sessionUpdates.add(keyIndex)
        }
    }

    fun binarySearch(key: K): Int = Collections.binarySearch(orderedKeys, key)

    fun stopSession() {
        sessionStarted = false
    }

    fun getKeyOrNull(index: Int): K? = orderedKeys.getOrNull(index)
    fun getValueOrNull(index: Int): V? = values.getOrNull(index)
    fun wasUpdated(index: Int) = sessionUpdates.contains(index)
    fun wasAdded(index: Int) = sessionInsertions.contains(index)
}

private enum class GLType(val niceName: String, val getterFunction: (Int) -> Any?) {
    BOOLEAN("boolean", GL11::glGetBoolean),
    FLOAT("float", GL11::glGetFloat),
    DOUBLE("double", GL11::glGetDouble),
    INTEGER("int", GL11::glGetInteger),
    INTEGER64("long", GL32::glGetInteger64);
}

private data class GLVariable(val pname: Int, val type: GLType) : Comparable<GLVariable> {
    override fun compareTo(other: GLVariable): Int =
        pname.compareTo(other.pname).takeUnless { it == 0 } ?: type.compareTo(other.type)

    fun getName(nicePName: String?): String = (nicePName ?: pname.toString()) + " [${type.niceName}]"
}
interface IOpenGLDebugger {
    var printOnCycle: Int
    fun start()
    fun stop()
    fun cycle()
}

class OpenGLDebugger(override var printOnCycle: Int) : IOpenGLDebugger {
    private val probeGUIThread: Thread
    private var cycleCount = 0
    private var started = false
    private val glStuff = ChangeTrackingSortedMap<GLVariable, Any>()
    private val glVariablesToCheck = ArrayList<GLVariable>()
    private val debuggerLock = ReentrantLock()
    private val glVariablesCondition = debuggerLock.newCondition()
    private var glVariablesUpdated = false

    private inner class SpecialTableModel : DefaultTableModel() {
        override fun isCellEditable(row: Int, column: Int): Boolean {
            return false
        }

        override fun getColumnName(column: Int): String {
            return when (column) {
                0 -> "Key"
                1 -> "Value"
                else -> ""
            }
        }

        override fun getRowCount(): Int {
            return glStuff.size
        }

        override fun getColumnCount(): Int {
            return 2
        }

        override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
            return glStuff.getKeyOrNull(rowIndex)?.let { glVariable ->
                when (columnIndex) {
                    0 -> glVariable.getName(niceToHave[glVariable.pname])
                    1 -> glStuff.getValueOrNull(rowIndex) ?: "<NULL>"
                    else -> ""
                }
            } ?: "<NOT FOUND>"
        }
    }

    private inner class SpecialTableCellRenderer : DefaultTableCellRenderer() {
        override fun getTableCellRendererComponent(
            table: JTable?,
            value: Any,
            isSelected: Boolean,
            hasFocus: Boolean,
            row: Int,
            column: Int
        ): Component {
            if (table == null) {
                return this
            }
            background = null
            super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column)
            if (glStuff.wasUpdated(row)) {
                background = Color.YELLOW
            } else if (glStuff.wasAdded(row)) {
                background = Color.GREEN
            }
            return this
        }
    }

    private inner class ProbeGUI : JDialog(), Runnable {
        private val mainPanel: JPanel
        private val variableMonitorTable: JTable
        private val closeButton: JButton

        init {
            for (info in UIManager.getInstalledLookAndFeels()) {
                if ("Windows" == info.name) {
                    try {
                        UIManager.setLookAndFeel(info.className)
                    } catch (e: ReflectiveOperationException) {
                        e.printStackTrace()
                    } catch (e: UnsupportedLookAndFeelException) {
                        e.printStackTrace()
                    }
                    break
                }
            }
            mainPanel = JPanel()
            closeButton = JButton().apply {
                text = "Close"
            }
            variableMonitorTable =
                JTable().apply {
                    isEnabled = false
                    model = SpecialTableModel()
                    columnModel.getColumn(0).cellRenderer = SpecialTableCellRenderer()
                }
            setupUI()
            type = Type.UTILITY
            contentPane = mainPanel
            getRootPane().defaultButton = closeButton
            defaultCloseOperation = DO_NOTHING_ON_CLOSE
            addWindowListener(object : WindowAdapter() {
                override fun windowClosing(e: WindowEvent) {
                    stop()
                }
            })
            closeButton.addActionListener { stop() }
            mainPanel.registerKeyboardAction(
                { stop() },
                KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0),
                JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT
            )
        }

        override fun run() {
            glVariablesToCheck.clear()
            glVariablesToCheck.addAll(
                niceToHave.keys.flatMap { key ->
                    GLType.values().map { glType -> GLVariable(key, glType) }
                }
            )
            pack()
            isVisible = true
            while (started) {
                debuggerLock.withLock {
                    try {
                        while (!glVariablesUpdated) {
                            glVariablesCondition.await()
                        }
                        glVariablesUpdated = false
                        updateVariables()
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                }
            }
        }

        private fun updateVariables() {
            val dtm = variableMonitorTable.model as DefaultTableModel
            dtm.fireTableDataChanged()
        }

        private fun setupUI() = mainPanel.apply {
            layout = GridBagLayout()
            border = BorderFactory.createTitledBorder(
                BorderFactory.createEmptyBorder(5, 5, 5, 5),
                "null",
            )

            add(
                closeButton,
                GridBagConstraints().apply {
                    gridx = 2
                    gridy = 1
                    fill = GridBagConstraints.HORIZONTAL
                    insets = Insets(0, 5, 0, 0)
                },
            )

            add(
                JTextField().apply {
                    text = "Search…"
                    // TODO: add functionality
                },
                GridBagConstraints().apply {
                    gridx = 0
                    gridy = 1
                    weightx = 1.0
                    anchor = GridBagConstraints.WEST
                    fill = GridBagConstraints.HORIZONTAL
                },
            )

            add(
                JScrollPane().apply {
                    horizontalScrollBarPolicy = ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER
                    setViewportView(variableMonitorTable)
                },
                GridBagConstraints().apply {
                    gridx = 0
                    gridy = 0
                    gridwidth = 3
                    weightx = 1.0
                    weighty = 1.0
                    fill = GridBagConstraints.BOTH
                    insets = Insets(0, 0, 5, 0)
                },
            )
        }
    }

    init {
        require(printOnCycle >= 1) { "Print per cycle must be at least 1" }
        probeGUIThread = Thread(
            /* target = */ {
                synchronized(niceToHave) {
                    if (niceToHave.isEmpty()) {
                        updateNiceToHave()
                    }
                }
                ProbeGUI().run()
            },
            /* name = */ "LogisticsPipes GLDebug Probe #$probeID",
        )
        probeID++
    }

    override fun start() {
        if (!started) {
            started = true
            cycleCount = 0
            probeGUIThread.start()
        }
    }

    override fun stop() {
        if (started) {
            debuggerLock.withLock {
                started = false
                glVariablesUpdated = true
                glVariablesCondition.signal()
            }
        }
    }

    override fun cycle() {
        if (!started) {
            start()
        }
        ++cycleCount
        if (cycleCount % printOnCycle == 0) {
            saveOpenGLStuff()
            cycleCount = 0
        }
    }

    private fun saveOpenGLStuff() {
        debuggerLock.withLock {
            glStuff.startSession()
            val toRemove = glVariablesToCheck.filter { variable ->
                LogisticsPipes.log.warn("Getting ${variable.getName(niceToHave[variable.pname])}")
                val value = variable.type.getterFunction(variable.pname)
                val glError = GL11.glGetError()
                val isInvalid = glError == GL11.GL_INVALID_ENUM
                if (glError != GL11.GL_NO_ERROR) {
                    LogisticsPipes.log.warn("  got error ${GLU.gluErrorString(glError)}")
                }
                value?.takeUnless { isInvalid }?.also {
                    glStuff.put(variable, it)
                }
                return@filter isInvalid
            }.toSet()
            glVariablesToCheck.removeAll(toRemove)
            glStuff.stopSession()
            glVariablesUpdated = true
            glVariablesCondition.signal()
        }
    }

    companion object {
        private val niceToHave: HashMap<Int, String> = HashMap()
        private var probeID = 0

        private fun updateNiceToHave() {
            niceToHave.clear()
            var crawlerVersion = 11
            var almostEnd = false
            var end = false
            while (!end) {
                val packageGL = String.format("%s%d", "GL", crawlerVersion)
                val nextGL = String.format("%s.%s", "org.lwjgl.opengl", packageGL)
                try {
                    crawlerVersion++
                    val glClass = GL11::class.java.classLoader.loadClass(nextGL)
                    @Suppress("UnstableApiUsage")
                    Reflection.initialize(glClass)
                    almostEnd = false
                    for (f in glClass.declaredFields) {
                        try {
                            if (f.type != Int::class.javaPrimitiveType) {
                                continue
                            }
                            val id = f.getInt(null)
                            if (id == GL31.GL_INVALID_INDEX) {
                                continue
                            }
                            val nice = f.name
                            if (nice.endsWith("BIT")) {
                                continue
                            }

                            // All the things that are being replaced are not that bad
                            if (niceToHave.containsKey(id) && niceToHave[id] != nice) {
                                System.out.printf(
                                    "NiceToHave: ID %d exists. Replacing %s with %s!!%n",
                                    id,
                                    niceToHave.remove(id),
                                    nice
                                )
                            }
                            niceToHave[id] = String.format("%s.%s", packageGL, nice)
                        } catch (e: IllegalArgumentException) {
                            System.out.printf("NiceToHave: Illegal Argument!%nNiceToHave: %s%n", e)
                            e.printStackTrace()
                        } catch (e: IllegalAccessException) {
                            System.out.printf("NiceToHave: Illegal Access!%nNiceToHave: %s%n", e)
                            e.printStackTrace()
                        }
                    }
                } catch (e: ClassNotFoundException) {
                    if (almostEnd) {
                        end = true
                    } else {
                        almostEnd = true
                        crawlerVersion = (crawlerVersion / 10 + 1) * 10
                    }
                }
            }
        }
    }
}

@Suppress("unused")
object GlobalOpenGLDebugger : IOpenGLDebugger by OpenGLDebugger(60)
